### PR TITLE
o/h/ctlcmd: support writing to registry views in snapctl

### PR DESCRIPTION
**Pull request by MiguelPires**
Adds registry support to `snapctl set` so snaps can write to registry views, if they have a matching connected plug for the view and if the registry assertion can be found locally.

### CODE REVIEW

**Review by andrewphelpsj — APPROVED** [review]
Thanks! One small comment.

File: overlord/hookstate/ctlcmd/set.go [hunk 1]
```diff
@@ -99,6 +102,27 @@ func (s *setCommand) Execute(args []string) error {
...
+			splitConf := strings.SplitN(conf, "=", 2)
+			if len(splitConf) == 1 && strings.HasSuffix(conf, "!") {
+				key := strings.TrimSuffix(conf, "!")
+				requests[key] = nil
+				continue
+			}
+
+			if len(splitConf) != 2 {
+				return fmt.Errorf("cannot set %s plug: must set field with \"=\" or unset with \"!\"", s.Positional.PlugOrSlotSpec)
+			}
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/set.go:118]: couldn't a helper be shared with the config code for some of this? otoh even if not shared maybe the loop/part the loop should be in a helper?

File: overlord/hookstate/ctlcmd/get.go [hunk 2]
```diff
@@ -354,34 +356,53 @@ func (c *getCommand) getInterfaceSetting(context *hookstate.Context, plugOrSlot
...
+
+	res, err := registrystate.GetViaViewInTx(tx, view, requests)
+	if err != nil {
+		return err
+	}
+
+	return c.printPatch(res)
+}
+
+func getRegistryView(ctx *hookstate.Context, plugName string) (*registry.Registry, *registry.View, error) {
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/get.go:384]: would it make sense to have a method on View that lets access the corresponding Registry to simplify things a bit here?

**Review by pedronis — COMMENTED** [review]
thank you, couple of comments

**Review by pedronis — COMMENTED** [review]
thx, code placement comment

File: cmd/snap/cmd_set.go [hunk 3]
```diff
@@ -26,7 +26,7 @@ import (
 	"github.com/jessevdk/go-flags"
 
 	"github.com/snapcore/snapd/i18n"
-	"github.com/snapcore/snapd/jsonutil"
+	"github.com/snapcore/snapd/overlord/hookstate/ctlcmd"
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply cmd/snap/cmd_set.go]: Thank you, I was looking for a place like this but didn't see an appropriate location. Changed this now

**Comment by MiguelPires** [comment]
The last commit fixes a bug caused by the refactor. The helper returns a map and ctlcmd/set.go iterated through it in any order so writes could be re-ordered relative to the input order unlike the original code. This feel like an implementation detail that no one should rely on (it also doesn't happen in `snap set`) but I'm wary of changing it in snapctl and potentially breaking snaps. This is also an issue for registries but it's easier to deal with if we split the requests from a map to two slices in registrystate. I can do that in a follow-up PR

**Review by pedronis — APPROVED** [review]
thank you

**Review by andrewphelpsj — APPROVED** [review]
this looks good, thanks!

**Pull request merged**